crates/bevy_openxr/src/interaction.rs

```diff
@@ -0,0 +1,493 @@
+pub const DAYDREAM_PROFILE: &str = "/interaction_profiles/google/daydream_controller";
+pub const VIVE_PROFILE: &str = "/interaction_profiles/htc/vive_controller";
+pub const VIVE_PRO_PROFILE: &str = "/interaction_profiles/htc/vive_pro";
+pub const WMR_PROFILE: &str = "/interaction_profiles/microsoft/motion_controller";
+pub const XBOX_PROFILE: &str = "/interaction_profiles/microsoft/xbox_controller";
+pub const GO_PROFILE: &str = "/interaction_profiles/oculus/go_controller";
+pub const OCULUS_TOUCH_PROFILE: &str = "/interaction_profiles/oculus/touch_controller";
+pub const VALVE_INDEX_PROFILE: &str = "/interaction_profiles/valve/index_controller";
+
+// Pose actions
+pub const LEFT_GRIP_POSE: &str = "bevy_left_grip_pose";
+pub const RIGHT_GRIP_POSE: &str = "bevy_right_grip_pose";
+pub const LEFT_AIM_POSE: &str = "bevy_left_aim_pose";
+pub const RIGHT_AIM_POSE: &str = "bevy_right_aim_pose";
+
+// Virtual controllers action names. The user can override the virtual controllers behavior by
+// submitting `OpenXrBindingDesc`s with these names.
+pub const MENU_CLICK: &str = "bevy_menu_click";
+pub const LEFT_PRIMARY_CLICK: &str = "bevy_left_primary_click";
+pub const RIGHT_PRIMARY_CLICK: &str = "bevy_right_primary_click";
+// ... todo: add the rest
+pub const LEFT_VIBRATION: &str = "bevy_left_vibration";
+pub const RIGHT_VIBRATION: &str = "bevy_right_vibration";
+
+pub(crate) struct ControllerPairInputActions {
+    menu_click: xr::Action<bool>,
+    left_hand: ControllerInputActions,
+    right_hand: ControllerInputActions,
+}
+
+struct ControllerInputActions {
+    primary_click: xr::Action<bool>,
+    primary_touch: xr::Action<bool>,
+    secondary_click: xr::Action<bool>,
+    secondary_touch: xr::Action<bool>,
+    trigger_value: xr::Action<f32>,
+    trigger_touch: xr::Action<bool>,
+    squeeze_value: xr::Action<f32>,
+    directional_value: xr::Action<xr::Vector2f>,
+    directional_click: xr::Action<bool>,
+    directional_touch: xr::Action<bool>,
+}
+
+pub(crate) struct ControllerPairOutputActions {
+    left_vibration: xr::Action<xr::Haptic>,
+    right_vibration: xr::Action<xr::Haptic>,
+}
+
+#[derive(Clone)]
+pub struct OpenXrActionPath {
+    pub profile: &'static str,
+    pub path: &'static str,
+}
+
+#[derive(Clone)]
+pub enum OpenXrActionType {
+    BinaryInput,
+    FloatInput,
+    Vector2Input,
+    PoseInput,
+    VibrationOutput,
+}
+
+#[derive(Clone)]
+pub struct OpenXrBindingDesc {
+    pub name: &'static str,
+    pub paths: Vec<OpenXrActionPath>,
+    pub action_type: OpenXrActionType,
+}
+
+#[derive(Clone)]
+pub(crate) enum XrAction {
+    BinaryInput(xr::Action<bool>),
+    FloatInput(xr::Action<f32>),
+    Vector2Input(xr::Action<xr::Vector2f>),
+    PoseInput(xr::Action<xr::Posef>),
+    VibrationOutput(xr::Action<xr::Haptic>),
+}
+
+impl XrAction {
+    fn unwrap_binary(self) -> xr::Action<bool> {
+        match self {
+            Self::BinaryInput(action) => action,
+            _ => panic!(),
+        }
+    }
```

Comment:
All of these `unwrap` functions seem to be a bit on an anit-pattern imo. 
It makes more sense for a user to manually `match` on the enum, and then panic themselves if they want to make sure the `XrAction` is a specific action type.

Reply:
Yes, this needs to be reworked.